### ISSUE TITLE
Remove Cipher dependencies in TLS

### ISSUE THREAD

**Issue by mpg** [Mbed-TLS/mbedtls#5797]
Now that all uses of Cipher for cryptographic computations have been removed from TLS when `USE_PSA` is enabled, it is desirable to also remove non-cryptographic uses, such as getting sizes from a cipher ID (often from a ciphersuite definition).

This task is to remove such uses and replace them with versions based on PSA or at least independent from `cipher.c`. Note: the helper function `mbedtls_ssl_cipher_to_psa()` can be useful for that.

- [x] Provide a definition of `mbedtls_ssl_ciphersuite_get_cipher_key_bitlen()` that avoids the cipher module when PSA is enabled. (Note: in the non-PSA case, we have an will keep a hard dependency on `CIPHER_C`, so the current `#if defined(MBEDTLS_CIPHER_C)` can be replaced with `#if !defined(MBEDTLS_USE_PSA_CRYPTO)` and then this item is about filling it the `USE_PSA` case.)
- [x] In `mbedtls_ssl_tls13_populate_transform()` make sure `mbedtls_cipher_info_from_type()` is only called when `USE_PSA` is disabled. The variable `cipher_info` should not exists when `USE_PSA` id enabled, and the first argument when calling `mbedtls_ssl_cipher_to_psa()` near the end of the function should be `ciphersuite_info->cipher`.
- [x] In `mbedtls_ssl_tls13_generate_handshake_keys()` and `mbedtls_ssl_tls13_generate_application_keys()`, avoid calling `mbedtls_cipher_info_from_type()` and compute the size another way (``mbedtls_ssl_cipher_to_psa()` can get the key size, and for the IV size a new helper function may be needed).

**Comment by mpg** [comment]
Note: the following script has been used in order to identify uses of Cipher functions when `USE_PSA` is enabled:
```sh
# avoid false positives when those are included
sed -i 's/#ifndef.*/#if 0/' include/mbedtls/cipher.h include/mbedtls/cmac.h
scripts/config.py full # enables USE_PSA too
for i in library/ssl*.c library/x509*.c; do gcc -Iinclude -E $i -o $i.i; done
egrep 'mbedtls_cipher[^ ]*\(' library/*.c.i
```
After this task, this script's output should be empty.

**Comment by mpg** [comment]
Trying to flesh out how it would work. In `mbedtls_ssl_transform`, instead of 
```c
#if defined(MBEDTLS_USE_PSA_CRYPTO)
    mbedtls_svc_key_id_t psa_key_enc;           /*!<  psa encryption key      */
    mbedtls_svc_key_id_t psa_key_dec;           /*!<  psa decryption key      */
    psa_algorithm_t psa_alg;                    /*!<  psa algorithm           */
#else
    mbedtls_cipher_context_t cipher_ctx_enc;    /*!<  encryption context      */
    mbedtls_cipher_context_t cipher_ctx_dec;    /*!<  decryption context      */
#endif /* MBEDTLS_USE_PSA_CRYPTO */
```
we'd have
```c
#if defined(MBEDTLS_USE_PSA_CRYPTO) || defined(MBEDTLS_SSL_PROTO_TLS1_3)
    mbedtls_svc_key_id_t psa_key_enc;           /*!<  psa encryption key      */
    mbedtls_svc_key_id_t psa_key_dec;           /*!<  psa decryption key      */
    psa_algorithm_t psa_alg;                    /*!<  psa algorithm           */
#endif
#if !defined(MBEDTLS_USE_PSA_CRYPTO) && defined(MBEDTLS_SSL_PROTO_TLS1_2)
    mbedtls_cipher_context_t cipher_ctx_enc;    /*!<  encryption context      */
    mbedtls_cipher_context_t cipher_ctx_dec;    /*!<  decryption context      */
#endif
```

Then in `mbedtls_ssl_encrypt_buf()` someting like:
```c
/* near the top */
    unsigned char use_psa = 0;
#if defined(MBEDTLS_USE_PSA_CRYPTO)
    use_psa = 1;
#endif 
#if defined(MBEDTLS_SSL_PROTO_TLS1_3)
    if( transform->tls_version == MBEDTLS_SSL_VERSION_TLS1_3 )
       use_psa = 1;
#endif
/* later */
#if defined(MBEDTLS_USE_PSA_CRYPTO) || defined(MBEDTLS_SSL_PROTO_TLS1_3)
    if( use_psa )
    {
        /* PSA-based version */
    }
#endif
#if !defined(MBEDTLS_USE_PSA_CRYPTO) && defined(MBEDTLS_SSL_PROTO_TLS1_2)
    if( !use_psa )
    {
        /* legacy non-PSA version */
    }
#endif
```
Finally, `mbedtls_ssl_tls12_populate_tranform()` would remain unchanged and `mbedtls_ssl_tls13_populate_tranform()` could be simplified by removing the non-PSA version.

**Comment by gilles-peskine-arm** [comment]
I think it would be cleaner to evolve TLS 1.3 to pure PSA. But I don't have a big picture view. @ronald-cron-arm What do you think? Should we force shared 1.2/1.3 code to also support PSA for 1.3 when `MBEDTLS_USE_PSA_CRYPTO` is off, or should some 1.3 code keep `#ifdef USE_PSA_CRYPTO` in places?

**Comment by mpg** [comment]
I think evolving TLS 1.3 to be pure PSA including record protection would:

- simplify what we can tell our users: "TLS 1.3 uses PSA period", as opposed to "TLS 1.3 uses PSA... except for the parts that depend on `USE_PSA`".
- simplify TLS 1.3 code (you'd never have an `#ifdef USE_PSA`)
- but make the code shared with 1.2 slightly more complex

**Comment by mpg** [comment]
Ok, when I wrote that TLS use PSA unconditionally, except for the parts that as common with TLS 1.2, which is only record protection, I was wrong: there's another part that's common with TLS 1.2, and that's the running handshake hash(es). That part can't be easily handled like I suggested for record protection above, because we need to start the hashes before the version has been negotiated (actually, before the handshake hash algorithm itself has been negotiated, that's why we are starting several of them).

It could be handled, in build with TLS 1.3 and without `USE_PSA`, by starting both the PSA and the non-PSA version of each hash, then killing one of them as soon as the version is know. But that seems costly in terms of resources, so I don't think we want to do it just for the sake of having a clearer story.

(Found that by `grep -n USE_PSA library/ssl_tls13*` in the latest version of #5804 - got matches in `mbedtls_ssl_tls13_populate_transform()` (record protection) as I expected, but also in `mbedtls_ssl_reset_transcript_for_hrr()` (running handshake hash) which I had overlooked.)
